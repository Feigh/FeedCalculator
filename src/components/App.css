.vertspace{
    margin-top: 10px;
}