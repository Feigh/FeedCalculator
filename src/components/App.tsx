import * as React from "react";
import "./App.css";
import Select from 'react-select'; // Fixades med "files": ["./src/index.tsx"] tsconfig.json??
import { StateManager } from "react-select/lib/stateManager";

interface IProps { }
interface IState { weightinput:string, rerinput:string, mealsinput:string, foodinput:string, mealsinputoption:string, foodinputoption:string, result:number}
const mealsList = [
{ value: 3, label: "3" },
{ value: 4, label: "4" },
{ value: 5, label: "5" },
{ value: 6, label: "6" }
];
const foodList = [
    { value: 1.09, label: "1.09 kcal/gram" }
    ];
// Ändra så att komponenten under value tar in oncalcfunctionen som retunerar en siffra
let ResultField = function(props :any){
    return <textarea id="result" key="result" className="md-textarea form-control vertspace" value={props.resultState} rows={3}></textarea>
}

let Button = function(props: any){
    return <button id="calculate" key="calculate" onClick={props.onClick}  className="btn vertspace" >Beräkna</button>
}

// onChange ska inte anropas om det användarna skriver in inte är ett nummer
export let InputField = function(props: any){
    return <div> 
        <span>{props.label}</span>
        <input id={props.keyName} key={props.keyName} value={props.inputValue} onChange={props.onChange} className="form-control vertspace" ></input>
    </div>
}

export let ComboBoxField = function(props: any){
    return <div> 
        <span>{props.label}</span>
        <Select options={props.ListOptions} id={props.keyName} key={props.keyName} value={props.selectedValue} onChange={props.onChange} className="vertspace" />
    </div>
}

export let InputArea = function(props:any){
    return <div className="row">
        <div  className="col">
            <InputField inputValue={props.inputState.weightinput} onChange={props.onWeightFieldChange} keyName="weightinput" label="Vikt(kg)" /> 
            <ComboBoxField selectedValue={props.inputState.mealsinputoption} ListOptions={mealsList} onChange={props.onMealsFieldChange} keyName="mealsinput" label="Måltider per dag" />            
        </div>
        <div  className="col">
            <InputField inputValue={props.inputState.rerinput} onChange={props.onCalorisFieldChange} keyName="rerinput" label="%RER" />
            <ComboBoxField selectedValue={props.inputState.foodinputoption} ListOptions={foodList} onChange={props.onFoodFieldChange} keyName="foodlistinput" label="Foder" />
        </div>
        </div>
}

export function FeedCalculation(weight: string, rer: string, meals: string, food: string){
    const reweight = (Number(weight) * 30)+70;
    const refeed = reweight/Number(food);
    const remeal = refeed/Number(meals);
    const procent = Number(rer)/100;
    return Math.round(remeal*(procent)* 10 ) / 10;
}

export default class App extends React.Component<IProps, IState> {
    constructor(props: any){
        super(props);
        this.state = {weightinput:"", rerinput: "", mealsinput: "", foodinput: "", foodinputoption: null, mealsinputoption: null, result: 0};
        this.onWeightFieldChange = this.onWeightFieldChange.bind(this);
        this.onRERFieldChange = this.onRERFieldChange.bind(this);
        this.onMealsFieldChange = this.onMealsFieldChange.bind(this);
        this.onCalculate = this.onCalculate.bind(this);
        this.onFoodFieldChange = this.onFoodFieldChange.bind(this);
    }
    onWeightFieldChange(event:React.FormEvent<HTMLInputElement>) {
        this.setState(({weightinput:event.currentTarget.value}));
    }
    onRERFieldChange(event:React.FormEvent<HTMLInputElement>) {
        this.setState(({rerinput:event.currentTarget.value}));
    }
    onMealsFieldChange(selectoption: any) {
        this.setState(({mealsinput: selectoption.value}));
        this.setState(({mealsinputoption: selectoption}));
    }
    onFoodFieldChange(selectoption: any) {
        this.setState(({foodinput:selectoption.value}));
        this.setState(({foodinputoption:selectoption}));
    }
    // gör om så att oncalculate retunerar ett nummer som sedan sätts i value
    onCalculate (event:React.FormEvent<HTMLInputElement>) {
        event.preventDefault();
        this.setState({result: FeedCalculation(this.state.weightinput, this.state.rerinput, this.state.mealsinput, this.state.foodinput)});
    }

    render() {
        return <div className='container'>

                    <h1>Mat Beräknare</h1>
                <form>                 
                    <InputArea inputState={this.state} 
                        onWeightFieldChange={this.onWeightFieldChange} onCalorisFieldChange={this.onRERFieldChange} onMealsFieldChange={this.onMealsFieldChange}
                        onFoodFieldChange={this.onFoodFieldChange} />
                    <Button onClick={this.onCalculate} />                 
                    <ResultField resultState={this.state.result} />
                </form>        
            </div>
    }
};