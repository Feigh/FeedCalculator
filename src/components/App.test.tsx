import * as React from "react";
import App, { ComboBoxField, FeedCalculation } from "./App";
import {InputField, InputArea} from "./App";
import {create } from "react-test-renderer";
import * as ReactTestRenderer from "react-test-renderer";
import * as  ShallowRenderer from "react-test-renderer/shallow";

// Test som kontrollerar att felmeddelandet dyker upp om en användare skriver in nått annat än nummer
// Test som kontrollerar att beräkningen ger rätt resultat
// Test som kontrollerar att InputField onchange inte anropas om värdet från användaren inte är siffra
// Är snapshots främst för regressions test, inte tdd?
// Jag renderar en komponent, hur kör jag dess functioner sedan?
// Generell, ska man testa state eller är det att missa ideen.
// Lägg till error span precis under label spann Testa att när komponentet inputtest skapas så syns inte texten
// 
describe("Test the App", () =>{
    it("should be container", () =>{
        const tree = create(<App/>).toJSON();
        expect(tree.props).toEqual({className: 'container'});
    })
    it("should change state", () =>{
        const renderer = create(<App/>);
        const instance = renderer.root;
        const component = instance.findByProps({className: 'container'})
    })
})

describe("Test the InputField", () =>{
    it("should return a container with props set container", () =>{
        const tree = create(<InputField keyName="testkey" label="testlbl"/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})

describe("Test the ComboBoxField", () =>{
    it("should return a container with props set container", () =>{
        const tree = create(<ComboBoxField keyName="testkey" label="testlbl"/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("should fil the optionslist", () =>{
        var list = [1,2,3];
        const tree = create(<ComboBoxField keyName="testkey" label="testlbl" value={list}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})

describe("Test the InputArea", () =>{
    it("should set Child component with right input state", () =>{
        let state = {weightinput:"vikt", rerinput: "rer", mealsinput: "meal", foodlistinput: "food"};
        const tree = create(<InputArea inputState={state} />).toJSON();
        expect(tree.children[0].children[1].props.value).toEqual("vikt");
        expect(tree.children[1].children[1].props.value).toEqual("rer");
        expect(tree.children[2].children[1].props.value).toEqual("meal");
        expect(tree.children[3].children[1].props.value).toEqual("food");
    })
})

describe("Test the Calculation", () =>{
    it("should return right result", () =>{
        const result =  FeedCalculation("20", "50", "4", "1.09");
        expect(result).toBe(76.8);
    })
})