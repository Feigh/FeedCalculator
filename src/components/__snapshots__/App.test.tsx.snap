// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the ComboBoxField should fil the optionslist 1`] = `
<div>
  <span>
    testlbl
  </span>
  <select
    className="form-control vertspace"
    id="testkey"
  />
</div>
`;

exports[`Test the ComboBoxField should return a container with props set container 1`] = `
<div>
  <span>
    testlbl
  </span>
  <select
    className="form-control vertspace"
    id="testkey"
  />
</div>
`;

exports[`Test the InputField should return a container with props set container 1`] = `
<div>
  <span>
    testlbl
  </span>
  <input
    className="form-control vertspace"
    id="testkey"
  />
</div>
`;
